File: jps/jps-builders/src/org/jetbrains/jps/incremental/storage/SourceToOutputMappingImpl.java
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.jps.incremental.storage;

import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.jps.builders.storage.SourceToOutputMapping;
import org.jetbrains.jps.incremental.relativizer.PathRelativizerService;

import java.io.IOException;
import java.nio.file.Path;
import java.util.Iterator;
import java.util.List;

@ApiStatus.Internal
public final class SourceToOutputMappingImpl implements SourceToOutputMapping, StorageOwner {
  private final OneToManyPathsMapping myMapping;

  public SourceToOutputMappingImpl(@NotNull Path storePath, @NotNull PathRelativizerService relativizer) throws IOException {
    myMapping = new OneToManyPathsMapping(storePath, relativizer);
  }

  @Override
  public void setOutputs(@NotNull String srcPath, @NotNull List<String> outputs) throws IOException {
    myMapping.setOutputs(srcPath, outputs);
  }

  @Override
  public void setOutput(@NotNull String srcPath, @NotNull String outputPath) throws IOException {
    myMapping.setOutput(srcPath, outputPath);
  }

  @Override
  public void appendOutput(@NotNull String srcPath, @NotNull String outputPath) throws IOException {
    myMapping.appendData(srcPath, outputPath);
  }

  @Override
  public void remove(@NotNull String srcPath) throws IOException {
    myMapping.remove(srcPath);
  }

  @Override
  public void removeOutput(@NotNull String srcPath, @NotNull String outputPath) throws IOException {
    myMapping.removeData(srcPath, outputPath);
  }

  @Override
  public @Nullable List<String> getOutputs(@NotNull String srcPath) throws IOException {
    return myMapping.getOutputs(srcPath);
  }

  @Override
  public @NotNull Iterator<String> getSourcesIterator() throws IOException {
    return myMapping.getKeysIterator();
  }

  @Override
  public @NotNull SourceToOutputMappingCursor cursor() throws IOException {
    return myMapping.cursor();
  }

  @Override
  public void flush(boolean memoryCachesOnly) {
    myMapping.flush(memoryCachesOnly);
  }

  @Override
  public void close() throws IOException {
    myMapping.close();
  }

  @Override
  public void clean() throws IOException {
    myMapping.clean();
  }
}

File: jps/jps-builders/src/org/jetbrains/jps/incremental/storage/SourceToOutputMappingImpl.kt
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.jps.incremental.storage

import org.jetbrains.annotations.ApiStatus
import org.jetbrains.jps.builders.storage.SourceToOutputMapping
import org.jetbrains.jps.incremental.relativizer.PathRelativizerService
import java.io.IOException
import java.nio.file.Path

@ApiStatus.Internal
class SourceToOutputMappingImpl @Throws(IOException::class) constructor(
  storePath: Path,
  relativizer: PathRelativizerService,
) : SourceToOutputMapping, StorageOwner {
  private val mapping = OneToManyPathsMapping(storePath, relativizer)

  override fun setOutputs(sourceFile: Path, outputs: List<Path>) {
    mapping.setOutputs(sourceFile, outputs)
  }

  override fun appendOutput(sourcePath: String, outputPath: String) {
    mapping.appendData(sourcePath, outputPath)
  }

  @Throws(IOException::class)
  override fun remove(sourceFile: Path) {
    mapping.remove(sourceFile)
  }

  @Throws(IOException::class)
  override fun removeOutput(srcPath: String, outputPath: String) {
    mapping.removeData(srcPath, outputPath)
  }

  @Throws(IOException::class)
  override fun getOutputs(srcPath: String): List<String>? = mapping.getOutputs(srcPath)

  @Throws(IOException::class)
  override fun getOutputs(sourceFile: Path): Collection<Path>? = mapping.getOutputs(sourceFile)

  @Throws(IOException::class)
  override fun getSourcesIterator(): Iterator<String> = mapping.keysIterator

  @Throws(IOException::class)
  override fun getSourceFileIterator(): Iterator<Path> {
    return mapping.keysIterator.asSequence().map { Path.of(it) }.iterator()
  }

  @Throws(IOException::class)
  override fun cursor(): SourceToOutputMappingCursor {
    return mapping.cursor()
  }

  override fun flush(memoryCachesOnly: Boolean) {
    mapping.flush(memoryCachesOnly)
  }

  override fun close() {
    mapping.close()
  }

  override fun clean() {
    mapping.clean()
  }
}
